File: tree/check_bst_contains_dead_end.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>

using namespace std;

struct Node {
  int data;
  struct Node *left;
  struct Node *right;

  Node(int x) {
    data = x;
    left = NULL;
    right = NULL;
  }
};

void insert(Node **tree, int val) {
  Node *temp = NULL;
  if (!(*tree)) {
    temp = new Node(val);
    *tree = temp;
    return;
  }

  if (val < (*tree)->data) {
    insert(&(*tree)->left, val);
  } else if (val > (*tree)->data) {
    insert(&(*tree)->right, val);
  }
}

int getCountOfNode(Node *root, int l, int h) {
  if (!root)
    return 0;

  if (root->data == h && root->data == l)
    return 1;

  if (root->data <= h && root->data >= l)
    return 1 + getCountOfNode(root->left, l, h) +
           getCountOfNode(root->right, l, h);

  else if (root->data < l)
    return getCountOfNode(root->right, l, h);

  else
    return getCountOfNode(root->left, l, h);
}

bool isDeadEnd(Node *root);
int main() {
  int T;
  cin >> T;
  while (T--) {
    Node *root;
    Node *tmp;
    // int i;

    root = NULL;

    int N;
    cin >> N;
    for (int i = 0; i < N; i++) {
      int k;
      cin >> k;
      insert(&root, k);
    }

    cout << isDeadEnd(root);
    cout << endl;
  }
}

// } Driver Code Ends

/*The Node structure is
struct Node {
int data;
Node * right, * left;
};*/

bool ok(Node *p, int mi, int mx) {
  if (!p)
    return false;

  // cout << p->data << ' ' << mi << ' ' << mx << endl;

  if (!p->left && !p->right) {
    if (p->data == mi + 1 && p->data == mx - 1)
      return true;
  }

  return ok(p->left, mi, p->data) || ok(p->right, p->data, mx);
}

/*You are required to complete below method */
bool isDeadEnd(Node *root) {
  // Your code here
  if (!root)
    return false;
  return ok(root, 0, INT_MAX);
}

File: tree/check_bst_contains_dead_end.py
# Your task is to complete this function
# function should return true/false or 1/0
class Solution:
    def isDeadEnd(self, root):
        # Code here
        def f(p, mi, ma):
            if p is None:
                return False
            if f(p.left, mi, p.data):
                return True
            if mi + 1 == p.data and ma - 1 == p.data:
                return True
            if f(p.right, p.data, ma):
                return True
            return False

        return f(root, 0, 20000)


# {
# Driver Code Starts


class Node:
    def __init__(self, value):
        self.left = None
        self.data = value
        self.right = None


def insert(root, node):
    if root is None:
        root = node
    else:
        if root.data < node.data:
            if root.right is None:
                root.right = node
            else:
                insert(root.right, node)
        elif root.data == node.data:
            return
        else:
            if root.left is None:
                root.left = node
            else:
                insert(root.left, node)


def traverseInorder(root):
    if root is not None:
        traverseInorder(root.left)
        print(root.data, end=" ")
        traverseInorder(root.right)


if __name__ == "__main__":
    t = int(input())
    for i in range(t):
        n = int(input())
        arr = list(map(int, input().strip().split()))
        root = None
        for j in arr:
            if root is None:
                root = Node(j)
            else:
                insert(root, Node(j))

        ob = Solution()
        if ob.isDeadEnd(root):
            print(1)
        else:
            print(0)
# Contributed by: Harshit Sidhwa

# } Driver Code Ends        for i in range(0, len(ans)-1):
